pkg/git/libgit2/transport.go

```diff
@@ -126,20 +128,24 @@ func (s *PublicKeyAuth) Method(secret corev1.Secret) (*git.Auth, error) {
 		user = git.DefaultPublicKeyAuthUser
 	}
 
-	credCallback := func(url string, username_from_url string, allowed_types git2go.CredType) (*git2go.Cred, error) {
+	credCallback := func(url string, usernameFromURL string, allowedTypes git2go.CredType) (*git2go.Cred, error) {
 		cred, err := git2go.NewCredSshKeyFromMemory(user, "", string(identity), "")
 		if err != nil {
 			return nil, err
 		}
 		return cred, nil
 	}
 	certCallback := func(cert *git2go.Certificate, valid bool, hostname string) git2go.ErrorCode {
+		host, _, err := net.SplitHostPort(s.host)
+		if err != nil || host != hostname {
+			return git2go.ErrUser
+		}
 		for _, k := range kk {
-			if k.matches(hostname, cert.Hostkey.HashSHA1[:]) {
+			if k.matches(s.host, cert.Hostkey.HashSHA1[:]) {
```

Comment:
This will compare (e.g.) `localhost:56456` with `[localhost]:56456` and fail when it should succeed. I think you want to construct a key to match with

```
if port != "22" { // i.e,. non-standard port
  host = "[" + host + "]:" + port
}
```

then use `host` in the comparison.

Reply:
I have changed it so that the `host` is formatted into `[%s]:%s` if `port != ""`, as a `known_hosts` entry with `[github.com]:22` would otherwise still not match.

Reply:
The entries in the known_hosts are created such that if the port is "22" or left out, it will be normalised to just the hostname. It occurs to me that the code in question here could just use the same procedure to generate a lookup value -- it's https://pkg.go.dev/golang.org/x/crypto/ssh/knownhosts#Normalize.

Reply:
:100: for pkg.go.dev/golang.org/x/crypto/ssh/knownhosts#Normalize